packages/lowerdash/src/types.ts

```diff
@@ -100,6 +100,11 @@ export const isNonEmptyArray = <T> (array: T[]): array is NonEmptyArray<T> => (
   array.length > 0
 )
 
+export const isTypeOfOrUndefined = <S>(
+  value: unknown,
+  typeGuard: TypeGuard<unknown, S>
+): value is S | undefined => value === undefined || typeGuard(value)
```

Comment:
shouldn't have to force the input type to be `unknown`
```suggestion
export const isTypeOfOrUndefined = <T, S extends T>(
  value: T,
  typeGuard: TypeGuard<T, S>
): value is S | undefined => value === undefined || typeGuard(value)
```

Reply:
ts fails with an error:
```
A type predicate's type must be assignable to its parameter's type.
  Type 'S | undefined' is not assignable to type 'T'.
    'T' could be instantiated with an arbitrary type which could be unrelated to 'S | undefined'.ts(2677)
```

Reply:
```typescript
export const isTypeOfOrUndefined = <T, S extends T>(
  value: T | undefined,
  typeGuard: TypeGuard<T, S>
): value is S | undefined => value === undefined || typeGuard(value)
```